### PR TITLE
ci for opening prs to upgrade and migrate nx

### PR DESCRIPTION
**Pull request by brossetti1**
## GitHub Issue

fixes https://github.com/HausDAO/daohaus-monorepo/issues/322

sample release branch (move to the bottom to see the releases - this is an exmaple run on this branch so none of the commits are in develop yet):

https://github.com/HausDAO/daohaus-monorepo/compare/bump-lib-versions-2022-08-10-095437?expand=1

(scroll down)

Screen Shot 2022-08-10 at 3 26 56 PM

## Changes

- [X] add https://github.com/jscutlery/semver
- [X] configure nx task for versioning
- [X] add `workflow_dispatch` task to generate versioned branch
- [X] default commit lint to not restrict commit messages off the bat, instead the lint rules that are required will show when you commit without a semantic version tag
- [X] added some caching to the ci build, probably needs a review

## Packages Added

```
"@commitlint/cli": "^17.0.0",
"@commitlint/config-conventional": "^17.0.0",
"@jscutlery/semver": "^2.26.0",
```

## Checks

Before making your PR, please check the following:

- [x] Critical lint errors are resolved
- [x] App runs locally
- [x] App builds locally (run the build command for *any impacted package* and check for any errors before the PR)

## Details

this branch adds semantic versioning to our monorepo libs that we currently have deployed to npm. the versioning system has features around:

1) calculates new version of each lib (based on tagging) 
2) creates/updates a changelog.md file (there might be some more things we can do here eventually, changelogs might not show alot of detail right now)
3) update package.json
4) adds the tag to the current version (will associate the tag with the latest commit)
5) creates a release

ex:
Screen Shot 2022-08-10 at 12 59 23 PM

# Semantic Versioning

ive defaulted the commit lint to **not** stop us from committing when we do not provide one of the tags for semantic versioning, instead you get a summary of the rules that would be broken if they were turned on:

Screen Shot 2022-08-10 at 12 20 55 PM

I directly added the config file for the [commitlint config - you can see these here](https://github.com/HausDAO/daohaus-monorepo/pull/522/files#diff-8f12d5f5467d5a9e7d05741bfaa3151879b70dc9b12ed2da94e21fcbbd80a986R1-R36)

there are a few prefixes you can use to allow semantic versioning to function:

```
# ["build","chore","ci","docs","feat","fix","perf","refactor","revert","style","test"]

> git commit -m 'feat: make feature work'
# bumps the minor version - 0.1.0

> git commit -m 'chore: i did a thing'
# versions ignored (unnecessary right now, until we turn commit lint on) - no version change

> git commit -m 'fix: I FIXED SOMETHING I BROKE'
# version bump 0.0.1 up
```

again, right now these are not linted, the rules are turned off.

# How to release

this is currently setup as a `workflow_dispatch` so when we are ready to merge develop into master, we can run this workflow manually, there are 2 ways to run it:

1) go to `Actions -> CI Release (left side) - at the top of the runs on right side, there is a button to dispatch the workflow and create the branch that bumps our packages, releases, and tags. 

https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch

2) call the workflow from the cli:

```
> gh workflow list

Ci branch   active  24259143
Ci branch   active  24259950
CI Release  active  32026227
CI Test     active  31780132
Nx migrate  active  31137376

> gh workflow run 'CI Release' --ref develop

✓ Created workflow_dispatch event for develop at develop
To see runs for this workflow, try: gh run list --workflow=ci_release.yml
```

**view ci tasks**
```
gh workflow view

? Select a workflow  [Use arrows to move, type to filter]
  Ci branch (ci_develop.yaml)
  Ci branch (ci_master.yaml)
> CI Release (ci_release.yml)
  CI Test (ci_test.yml)
  Nx migrate (nx-migrate.yml)

# hit enter

CI Release - ci_release.yml
ID: 32026227

Total runs 12
Recent runs
✓  removing pr open command, add commitlint  CI Release  ci-add-nrwl-upgrade  workflow_dispatch  2834400559
X  chore: login                              CI Release  ci-add-nrwl-upgrade  workflow_dispatch  2828945616
X  chore: i hate ci                          CI Release  ci-add-nrwl-upgrade  workflow_dispatch  2828693850
X  chore: gdi                                CI Release  ci-add-nrwl-upgrade  workflow_dispatch  2828688499
X  chore: freaking permissions?!@?           CI Release  ci-add-nrwl-upgrade  workflow_dispatch  2828677978

To see more runs for this workflow, try: gh run list --workflow ci_release.yml
To see the YAML for this workflow, try: gh workflow view ci_release.yml --yaml
```

# Other

heres what [caching looks like when multiple jobs are run on the same push](https://github.com/HausDAO/daohaus-monorepo/pull/522/files#diff-0e0ddb566e48059418751ede2c1cfea5cefc345b50465ad23aa1bb5188241970R23-R34)

Screen Shot 2022-08-10 at 3 58 34 PM

### CODE REVIEW

**Comment by alexkeating** [comment]
I have an nx migration in my pr. So, there might be some conflicts if mine gets in

**Comment by brossetti1** [comment]
@alexkeating yeah i saw that, no worries, i cna rebase. mainly trying to play around with devops stuff here -- looking into https://github.com/jscutlery/semver with this branch

**Review by 0xRowdy — APPROVED** [review]

File: .commitlintrc.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        "fix",
+        "perf",
+        "refactor",
+        "revert",
+        "style",
+        "test"
+      ]
+    ]
+  }
+}
```
Comment (alexkeating) [inline .commitlintrc.json:36]: Nice, I haven't heard of this

**Review by alexkeating — APPROVED** [review]
Looks good and the pr description is great. It would be good to migrate to the wiki or a readme

File: .github/workflows/ci_master.yaml [hunk 2]
```diff
@@ -1,15 +1,17 @@
 name: Ci branch
 
+# for testing purposes, should not go in pr;
+# switch to master instead of develop
 on:
   pull_request:
     branches:
-      - 'master'
+      - 'develop'
```
[the remark this answers is not in the document]
Comment (brossetti1) [reply .github/workflows/ci_master.yaml]: ah this was a typo - i was switching the release.yml to develop and must of switched to the master yml accidentally, thanks!

**Review by skuhlmann — APPROVED** [review]

**Pull request merged**